### PR TITLE
Add a handler for Jakarta Persistence to reason about field initialization

### PR DESCRIPTION
**Pull request by msridhar**
Detects when a field is externally initialized via JPA persistence, so we can skip initialization checking of such fields and avoid false positives.

## Summary by CodeRabbit

* **New Features**
  * Added Jakarta Persistence (JPA) support so null-initialization checks respect JPA access strategies and distinguish persistent vs non-persistent fields.

* **Improvements**
  * Handler-based skip logic unified for more consistent behavior.

* **Tests**
  * Added tests covering field/property access, inheritance, transient/static exclusions, and mapping edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/uber/NullAway/pull/1584?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by msridhar** [comment]
This change is part of the following stack:

- #1584 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java [hunk 1]
```diff
@@ -493,14 +492,17 @@ default Type.MethodType onOverrideMethodType(
...
-   * @param annotationMirror the annotation
+   * @param classSymbol the symbol for the class declaring the field
+   * @param fieldSymbol the symbol for the field
+   * @param state visitor state
    * @return {@code true} if initialization checking for the field should be skipped, {@code false}
    *     otherwise
    */
-  default boolean shouldSkipFieldInitializationCheck(AnnotationMirror annotationMirror) {
+  default boolean shouldSkipFieldInitializationCheck(
+      Symbol.ClassSymbol classSymbol, Symbol fieldSymbol, VisitorState state) {
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/Handler.java:505]: Generally, NullAway does not have downstream clients that use it as a library, so we don't care too much about maintaining a stable public API

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+      return false;
+    }
+    // if the field itself has an @Access(FIELD) annotation, it is managed
+    if (hasJpaAccessAnnotation(fieldSymbol, "FIELD")) {
+      return true;
+    }
+    // otherwise, determine the access type for the enclosing class, and proceed appropriately
+    return switch (getJpaAccess(classSymbol)) {
+      case FIELD -> true;
+      case PROPERTY -> isBackingFieldForPersistentProperty(classSymbol, fieldSymbol);
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:167]: I think `@Access(PROPERTY)` annotations belong on accessor methods, not fields

**Review by yuxincs — APPROVED** [review]

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+    if (!methodSymbol.params().isEmpty() || methodSymbol.isStatic()) {
+      return null;
+    }
+    String methodName = methodSymbol.getSimpleName().toString();
+    if (methodName.startsWith("get") && methodName.length() > 3) {
+      String propertyName = methodName.substring(3);
+      return decapitalize(propertyName);
+    }
+    return null;
+  }
```
[the remark this answers is not in the document]
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:276]: I believe `is` getters are only for primitive `boolean` properties, which NullAway doesn't care about (since it's not a reference type)

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+  private static boolean isGetter(Symbol.MethodSymbol methodSymbol) {
+    return propertyNameForGetter(methodSymbol) != null;
+  }
+
+  private static @Nullable String propertyNameForGetter(Symbol.MethodSymbol methodSymbol) {
+    if (!methodSymbol.params().isEmpty() || methodSymbol.isStatic()) {
+      return null;
+    }
+    String methodName = methodSymbol.getSimpleName().toString();
+    if (methodName.startsWith("get") && methodName.length() > 3) {
```
Comment (askramana) [inline nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:267]: The Java Beans spec (and JPA property access) recognizes `isXxx()` as a valid getter for boolean properties.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+  /**
+   * For a field to be a backing field for a persistent property, there must be a corresponding
+   * getter and setter, and the getter should not be marked as transient
+   */
+  private static boolean isBackingFieldForPersistentProperty(
+      Symbol.ClassSymbol classSymbol, Symbol fieldSymbol) {
+    String propertyName = fieldSymbol.getSimpleName().toString();
+    Symbol.MethodSymbol getter = null;
+    boolean hasSetter = false;
+    for (Symbol member : classSymbol.members().getSymbols()) {
```
Comment (askramana) [inline nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:236]: `Symbol.members().getSymbols()` misses on the inherited ones. We need to walk up the hierarchy.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+  /** Annotations that explicitly select field-based or property-based persistence access. */
+  private static final Set<String> JPA_ACCESS_ANNOTS =
+      Set.of("javax.persistence.Access", "jakarta.persistence.Access");
+
+  /*
+   * JPA defaults an entity hierarchy to field access or property access based on where the mapping
+   * annotations are placed. We use this set to infer that access strategy when there is no
+   * explicit @Access annotation.
+   */
+  private static final Set<String> JPA_MAPPING_ANNOTS =
```
Comment (askramana) [inline nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:57]: Do we need to handpick these annotations? If it's desired to support all JPA annotations, below can work.

```java
private static boolean isJpaMappingAnnotation(String annotFqn) {
    return annotFqn.startsWith("javax.persistence.")
        || annotFqn.startsWith("jakarta.persistence.");
}
```

Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:57]: Some annotations in that namespace are not "mapping" annotations, e.g., `jakarta.persistence.Transient`, so I think we should be careful.  I _think_ all these annotations relate to how a field / property is mapped to the database (the list was generated by an LLM)?  In any case, for now I'd prefer to keep this list, but if it gets out of hand, we can approximate.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java (hunk 5 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:236]: Hrm, are you saying there can be a field `foo` where the corresponding `getFoo` and `setFoo` methods are in a superclass?  That seems like a bit strange / error-prone use of inheritance.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java (hunk 4 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:267]: Yes, at first I thought we could ignore this because NullAway doesn't care about `boolean` primitive-type fields.  But, it may be relevant for detecting the access mode for a class; so I will add it.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+ * not require explicit initialization.
+ *
+ * @see <a
+ *     href="https://jakarta.ee/learn/docs/jakartaee-tutorial/current/persist/persistence-intro/persistence-intro.html">Jakarta
+ *     Persistence documentation</a>.
+ */
+public class JakartaPersistenceHandler implements Handler {
+
+  /** Type annotations that mark a class as managed by JPA or Jakarta Persistence. */
+  private static final Set<String> JPA_MANAGED_TYPE_ANNOTS =
```
Comment (askramana) [inline nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:31]: We might also need to add `@IdClass`, `@Converter` to the list.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+      Set.of(
+          "javax.persistence.Entity",
+          "javax.persistence.MappedSuperclass",
+          "javax.persistence.Embeddable",
+          "jakarta.persistence.Entity",
+          "jakarta.persistence.MappedSuperclass",
+          "jakarta.persistence.Embeddable");
+
+  /** Field or accessor annotations that exclude a member from persistence handling. */
+  private static final Set<String> JPA_TRANSIENT_ANNOTS =
```
Comment (askramana) [inline nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:41]: Minor: Add `java.beans.Transient`. Found 1 reference for it within fievel.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java (hunk 5 above)
Comment (askramana) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:236]: It's normal with `@MappedSuperclass`.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java (hunk 7 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:31]: According to ChatGPT, these annotations are not equivalent to annotating a class with `@Entity` or the like, in terms of persistence / field initialization: [transcript](https://chatgpt.com/share/6a170a77-02e0-83e8-be8b-9e8a2de9482f).  Looking over the Javadoc for the annotations, I'm not seeing anything indicating they should be added.  But, will keep an eye out.

File: nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java (hunk 5 above)
Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:236]: Do you have an example?  If `getFoo` and `setFoo` are not abstract in the superclass, then they should be accessing their own backing field, not one in a subclass.  If `getFoo` and `setFoo` are abstract in the superclass, then they need to be overridden and will appear in the subclass directly.

Comment (askramana) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:236]: I don't need to override the subclass methods if there's no intention to modify their behavior.

```
@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
@lombok.Getter 
@lombok.Setter
@SuppressWarnings("NullAway.Init") // JPA requires no-args constructor
public abstract class AuditableEntity implements Serializable {

  private static final long serialVersionUID = 4949472584774822594L;

  @CreatedDate private LocalDateTime createdAt;
  @LastModifiedDate private LocalDateTime updatedAt;
}
```

```
@Entity 
@Table
@lombok.Getter 
@lombok.Setter
@SuppressWarnings("NullAway.Init") // JPA requires no-args constructor
public class User extends AuditableEntity {

  private static final long serialVersionUID = -6786279450970324630L;

  @Id
  @GeneratedValue(strategy = GenerationType.IDENTITY)
  @Nullable
  private Integer id;

  @Column
  private String name;
  ...
}
```

Comment (msridhar) [reply nullaway/src/main/java/com/uber/nullaway/handlers/JakartaPersistenceHandler.java:236]: I think this will work fine as there are generated getters and setters in both the superclass and subclass.  Note this routine is trying to determine if a particular field is a backing property, not the overall access mode for the class.  I should write a test for this though, including Lombok usage; will do.
